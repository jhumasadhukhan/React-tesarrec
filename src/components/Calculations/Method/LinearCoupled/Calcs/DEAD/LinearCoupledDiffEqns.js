import React from "react";
import MyChart from "../../../../../UI/Canvas/Charts/Chart";

import { parse, simplify, ConditionalNodeDependencies } from "mathjs";

//DyByDx= x+3y

const LinearCoupledDiffFourEqn = (props) => {
  /**
   * A method component that takes in a equation and outputs a chart
   */

  let parsedEquations = [];
  for (let i = 0; i < props.eqns.length; i++) {
    const eqn = props.eqns[i];
    var parsed = simplify(parse(eqn));
    parsedEquations.push(parsed);
  }

  const calcValueAt = (initialValues) => {
    // { 1 , 2, 3}
    let eqnResultsArr = [];
    let coordinate = { t: 0.1 }; //t initial value

    //generates the eval object
    for (let i = 0; i < props.LineNames.length; i++) {
      const dependentVariable = props.LineNames[i]; // { a, b , c}
      coordinate[dependentVariable] = initialValues[i];
    }

    for (let idx = 0; idx < parsedEquations.length; idx++) {
      eqnResultsArr.push(parsedEquations[idx].evaluate(coordinate)); // { a :1 , b: 3.3}
    }

    return eqnResultsArr;
  };

  const SolveLinearCoupledDifferentialEquationRungeKuttaForRth = (
    initialValues
  ) => {
    let eqnResults1_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults1_.push(initialValues[i]);
    }
    let eqnResults1 = calcValueAt(eqnResults1_);

    let eqnResults2_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults2_.push(initialValues[i] + (h / 2) * eqnResults1[i]);
    }
    let eqnResults2 = calcValueAt(eqnResults2_);

    let eqnResults3_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults3_.push(initialValues[i] + (h / 2) * eqnResults2[i]);
    }
    let eqnResults3 = calcValueAt(eqnResults3_);

    let eqnResults4_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults4_.push(initialValues[i] + (h / 2) * eqnResults3[i]);
    }
    let eqnResults4 = calcValueAt(eqnResults4_);

    let newY1Y2Y3Arr = [];

    for (let i = 0; i < props.LineNames.length; i++) {
      newY1Y2Y3Arr.push(
        initialValues[i] +
          h *
            (eqnResults1[i] / 6 +
              eqnResults2[i] / 3 +
              eqnResults3[i] / 3 +
              eqnResults4[i] / 6)
      );
    }

    return newY1Y2Y3Arr;
  };

  const recursive = (numberOfLoops, Arr) => {
    let allArrs = [abcArr];
    for (let i = 0; i < numberOfLoops; i++) {
      Arr = SolveLinearCoupledDifferentialEquationRungeKuttaForRth(Arr);
      allArrs.push(Arr);
    }
    console.log(allArrs);
    
    return allArrs;
  };

  const ShowLinearCoupledGraph = (EqnArr) => {


    return (
      <div>
        <div>
          <MyChart
            EulerData={EqnArr[0]}
            MidpointData={EqnArr[1]}
            RungeKuttaData={EqnArr[2]}
            Line4Data={EqnArr[3]}
            LineNames={props.LineNames}
            axisNames={["t", ""]}
            horizontalAlign={props.LegendHorizontal}
            verticalAlign={props.LegendVertical}
          />
        </div>
      </div>
    );
  };

  const FormatArrayLinearCoupled = (arr) => {
    let returnedArr = [];
    for (let j = 0; j < props.LineNames.length; j++) {
      let EqnArr = [];
      for (let i = 0; i < arr.length; i++) {
        const element = arr[i];
        EqnArr.push({
          x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
          y: parseFloat(element[j].toFixed(props.DecimalPrecision)),
        });
      }
      returnedArr.push(EqnArr);
    }
    return returnedArr;
  };

  // starts at x0=0

  let h = parseFloat(props.h);

  let abcArr = props.initialConditions;

  let CompletedGridRungeKuttaLinearCoupled = [];

  //SolveForRth(SolveForRth(Y0,h,X0))
  CompletedGridRungeKuttaLinearCoupled.push(
    recursive(props.numberOfCycles, abcArr)
  );

  let EqnArr = FormatArrayLinearCoupled(
    CompletedGridRungeKuttaLinearCoupled[0]
  );
  //return null;

  return ShowLinearCoupledGraph(EqnArr);
};

export default LinearCoupledDiffFourEqn;
