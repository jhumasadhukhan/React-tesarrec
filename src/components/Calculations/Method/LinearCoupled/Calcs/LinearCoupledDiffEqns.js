import React from "react";
import MyChart from "../../../../UI/Canvas/Charts/Chart";

import { parse, simplify } from "mathjs";

//DyByDx= x+3y

const LinearCoupledDiffFourEqn = (props) => {
  /**
   * A method component that takes in a equation and outputs a chart
   */

  
  let parsedEquations=[]
  for (let i = 0; i < props.eqns.length; i++) {
    const eqn = props.eqns[i];
    var parsed = simplify(parse(eqn));
    parsedEquations.push(parsed)
  }

  const calcValueAt = (initialValues) => { // { 1 , 2, 3}
    let eqnResultsArr = [];
    let coordinate = {t:1}; //t initial value
    
    //generates the eval object
    for (let i = 0; i < props.LineNames.length; i++) {
      const dependentVariable = props.LineNames[i]; // { a, b , c}
      coordinate[dependentVariable] = initialValues[i];
    }
    
    for (let idx = 0; idx < parsedEquations.length; idx++) {
      eqnResultsArr.push(parsedEquations[idx].evaluate(coordinate));// { a :1 , b: 3.3}
    }
    
    return eqnResultsArr;
  };

  const SolveLinearCoupledDifferentialEquationRungeKuttaForRth = (
    initialValues
  ) => {

    //var varCoords = {};
    //props.LineNames.forEach(function(item, index) {
    //  varCoords[item] = initialValues[index]

    //});

    let eqnResults1_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults1_.push(initialValues[i]);
    }
    let eqnResults1 = calcValueAt(eqnResults1_);

    let eqnResults2_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults2_.push(initialValues[i] + (h / 2) * eqnResults1[i]);
    }
    let eqnResults2 = calcValueAt(eqnResults2_);

    let eqnResults3_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults3_.push(initialValues[i] + (h / 2) * eqnResults2[i]);
    }
    let eqnResults3 = calcValueAt(eqnResults3_);

    let eqnResults4_ = [];
    for (let i = 0; i < props.LineNames.length; i++) {
      eqnResults4_.push(initialValues[i] + (h / 2) * eqnResults3[i]);
    }
    let eqnResults4 = calcValueAt(eqnResults4_);

    let newY1Y2Y3Arr = [];

    for (let i = 0; i < props.LineNames.length; i++) {
      newY1Y2Y3Arr.push(
        initialValues[i] +
          h *
            (eqnResults1[i] / 6 +
              eqnResults2[i] / 3 +
              eqnResults3[i] / 3 +
              eqnResults4[i] / 6)
      );
    }

    return newY1Y2Y3Arr;
  };

  const recursive = (numberOfLoops, Arr) => {

    let allArrs = [abcArr];
    for (let i = 0; i < numberOfLoops; i++) {
      Arr = SolveLinearCoupledDifferentialEquationRungeKuttaForRth(Arr);
      allArrs.push(Arr);
    }

    return allArrs;

    // for (let i = 0; i < numberOfLoops; i++) {
    //   if (method === "RungeKuttaLinearCoupled") {
    //     //   console.log(Arr)

    //     allArrs.push(Arr);
    //     return SolveLinearCoupledDifferentialEquationRungeKuttaForRth(Arr);
    //   }
    // }
  };

  const ShowLinearCoupledGraph = (EqnArr) => {
    // let DataNames=["EulerData", "MidpointData","RungeKuttaData","Line4Data"]

    // for (let i = 0; i < EqnArr.length; i++) {
    //     const Eqn = EqnArr[i];

    // }

    return (
      <div>
        <div>
          <MyChart
            EulerData={EqnArr[0]}
            MidpointData={EqnArr[1]}
            RungeKuttaData={EqnArr[2]}
            Line4Data={EqnArr[3]}
            LineNames={props.LineNames}
            axisNames={["t", ""]}
            horizontalAlign={props.LegendHorizontal}
            verticalAlign={props.LegendVertical}
          />
        </div>
      </div>
    );
  };

  const FormatArrayLinearCoupled = (arr) => {
    // let FirstEqnArr = [];
    // let SecondEqnArr = [];

    // for (let i = 0; i < arr.length; i++) {
    //   const element = arr[i];
    //   FirstEqnArr.push({
    //     x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
    //     y: parseFloat(element[0].toFixed(props.DecimalPrecision)),
    //   });
    //   SecondEqnArr.push({
    //     x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
    //     y: parseFloat(element[1].toFixed(props.DecimalPrecision)),
    //   });
    if (props.LineNames.length === 2) {
      let FirstEqnArr = [];
      let SecondEqnArr = [];

      for (let i = 0; i < arr.length; i++) {
        const element = arr[i];
        FirstEqnArr.push({
          x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
          y: parseFloat((element[0]).toFixed(props.DecimalPrecision)),
        });
        SecondEqnArr.push({
          x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
          y: parseFloat(element[1].toFixed(props.DecimalPrecision)),
        });
      }
      return [FirstEqnArr,SecondEqnArr];

    } else if (props.LineNames.length === 3) {
        let FirstEqnArr = [];
        let SecondEqnArr = [];
        let ThirdEqnArr = [];
    
        for (let i = 0; i < arr.length; i++) {
          const element = arr[i];
          FirstEqnArr.push({
            x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
            y: parseFloat(element[0].toFixed(props.DecimalPrecision)),
          });
          SecondEqnArr.push({
            x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
            y: parseFloat(element[1].toFixed(props.DecimalPrecision)),
          });
          ThirdEqnArr.push({
            x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
            y: parseFloat(element[2].toFixed(props.DecimalPrecision)),
          });
        }
        return [FirstEqnArr, SecondEqnArr, ThirdEqnArr];

    }else if (props.LineNames.length === 4) {
        let FirstEqnArr = [];
        let SecondEqnArr = [];
        let ThirdEqnArr = [];
        let FourthEqnArr = [];
    
        for (let i = 0; i < arr.length; i++) {
          const element = arr[i];
          FirstEqnArr.push({
            x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
            y: parseFloat(element[0].toFixed(props.DecimalPrecision)),
          });
          SecondEqnArr.push({
            x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
            y: parseFloat(element[1].toFixed(props.DecimalPrecision)),
          });
          ThirdEqnArr.push({
            x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
            y: parseFloat(element[2].toFixed(props.DecimalPrecision)),
          });
          FourthEqnArr.push({
            x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
            y: parseFloat(element[3].toFixed(props.DecimalPrecision)),
          });
        }
        return [FirstEqnArr, SecondEqnArr, ThirdEqnArr, FourthEqnArr];

    }

    // let allEqnsArr = [];
    // let EqnArr = [];
    // console.log(arr);

    // for (let i = 0; i < arr.length; i++) {
    //   const element = arr[i];
    //   for (let j = 0; j < 2; j++) {
    //     EqnArr.push({
    //       x: parseFloat((i * h).toFixed(props.DecimalPrecision)),
    //       y: parseFloat(element[i][j].toFixed(props.DecimalPrecision)),
    //     });
    //     allEqnsArr.push(EqnArr);
    //   }
    // }

    // for (let j = 0; j < arr.length; j++) {
    //   let EqnArr = [];

    //   for (let i = 0; i < props.LineNames.length; i++) {
    //     const element = arr[j];
    //     EqnArr.push({
    //       x: parseFloat((j * h).toFixed(props.DecimalPrecision)),
    //       y: parseFloat(element[i].toFixed(props.DecimalPrecision)),
    //     });
    //   }
    //   allEqnsArr.push(EqnArr);
    // }

  };

  // starts at x0=0

  let h = parseFloat(props.h);

  let abcArr = props.initialConditions;

  let CompletedGridRungeKuttaLinearCoupled = [];

  //SolveForRth(SolveForRth(Y0,h,X0))
  CompletedGridRungeKuttaLinearCoupled.push(recursive(props.numberOfCycles, abcArr));

  let EqnArr = FormatArrayLinearCoupled(CompletedGridRungeKuttaLinearCoupled[0]);
  //return null;

  return ShowLinearCoupledGraph(EqnArr);
};

export default LinearCoupledDiffFourEqn;
