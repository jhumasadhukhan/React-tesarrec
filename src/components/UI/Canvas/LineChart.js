import React, { Component } from "react";
//import classes from './Chart.module.css';
//import FractionDisplay from '../../Math/Math'
import {
  LineChart,
  Line,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
  Legend,
  ResponsiveContainer,
  Label,
} from "recharts";

// var CanvasJSChart = CanvasJSReact.CanvasJSChart;

// class LineChart extends Component {
//   render() {
//     let data = [];
//     for (let i = 0; i < this.props.dataPoints.length; i++) {
//       data.push({
//         type: "line",
//         // name: this.props.LineNames[i],

//         toolTipContent: " {x}, {y}",
//         dataPoints: this.props.dataPoints[i],
//         markerType: "none",
//       });
//     }

//     const options = {
//       animationEnabled: true,
//       // exportEnabled: true,
//       theme: "light2", // "light1", "dark1", "dark2"

//       title: {
//         text: this.props.title,
//         // more attributes
//       },
//       axisY: {
//         title: this.props.axisNames[1],
//       },
//       axisX: {
//         title: this.props.axisNames[0],
//       },
//       legend: {
//         verticalAlign: this.props.verticalAlign,
//         horizontalAlign: this.props.horizontalAlign,
//       },
//       backgroundColor: "white",
//       zoomEnabled: true,
//       data: data,
//     };
//     return (
//       <div style={{ width: "100%" }}>
//         <CanvasJSChart
//           options={options}
//           /* onRef={ref => this.chart = ref} */
//         />
//         {/*You can get reference to the chart instance as shown above using onRef. This allows you to access all chart properties and methods*/}
//       </div>
//     );
//   }
// }

class MyLineChart extends Component {
  render() {
    
    console.log(this.props.dataPoints[0][5].x)
    console.log(this.props.dataPoints[0][5].y)
    return (
      <div
        style={{
          paddingBottom: "56.25%" /* 16:9 */,
          position: "relative",
          height: 0,
        }}
      >
        <div
          style={{
            position: "absolute",
            top: "0",
            left: "0",
            width: "100%",
            height: "100%",
          }}
        >
          <ResponsiveContainer width="95%" height="100%">
            <LineChart
              data={this.props.dataPoints[0]}
              // margin={{ top: 5, right: 30, left: 20, bottom: 5 }}
            >
              <CartesianGrid strokeDasharray="3 3" />
              <XAxis
                dataKey="x"
                label={{ value: this.props.axisNames[0], position: "insideBottom" }}
              />

              <YAxis>
                <Label
                  value={this.props.axisNames[1]}
                  position="insideLeft"
                  angle={-90}
                  style={{ textAnchor: "middle" }}
                />
              </YAxis>
              <Tooltip />
              <Legend />
              <Line
                type="monotone"
                dataKey="y"
                stroke="#8884d8"
                legendType="none"
              />
              {/* <Line type="monotone" dataKey="y" stroke="#82ca9d" /> */}
            </LineChart>
          </ResponsiveContainer>
        </div>
      </div>
    );
  }
}

export default MyLineChart;
