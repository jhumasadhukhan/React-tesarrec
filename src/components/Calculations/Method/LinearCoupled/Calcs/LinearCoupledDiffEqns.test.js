import React from "react" ;
import LineChart from "../../../../UI/Canvas/Charts/Chart"
import {configure, shallow} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import LinearCoupledDiffEqns from './LinearCoupledDiffEqns'

configure({adapter:new Adapter()})
describe('<LinearCoupledDiffEqns/>', ()=>{
    it('should solve on differential eqn and presnet a chart', ()=>{

        let comp =  <LinearCoupledDiffEqns
        h={0.05}
        numberOfCycles={31}
        eqns={["a", "b"]}
        LineNames={["a", "b"]}
        initialConditions={[1,2]}
        LegendVertical={"center"}
        LegendHorizontal={"top"}
        DecimalPrecision={2}
      />
        const wrapper = shallow(comp)
        
        expect(wrapper.find(<LineChart/>))
        expect(wrapper.find('LineChart').props().LineNames.length).toEqual(2);
        expect(wrapper.find('LineChart').props().EulerData.length).toEqual(32);
        expect(wrapper.find('LineChart').props().MidpointData.length).toEqual(32);
        expect(wrapper.find('LineChart').props().MidpointData[0].y).toEqual(2);
        expect(wrapper.find('LineChart').props().EulerData[0].y).toEqual(1);

        //expect(wrapper.find(<LineChart/>).type()).to.equal('LineChart');
        //expect(wrapper.props().numberOfCycles).to.equal([ 'Euler', 'Midpoint', 'Runge Kutta' ]);
        
       
        //console.log(wrapper.find('EulerData').debug({ verbose: true }));
        //console.log(wrapper.debug({ verbose: true }));
        expect(wrapper).toMatchSnapshot();

    })
})

