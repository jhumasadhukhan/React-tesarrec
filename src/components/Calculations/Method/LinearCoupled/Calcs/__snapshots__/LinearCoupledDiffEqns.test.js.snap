// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinearCoupledDiffEqns/> should solve on differential eqn and presnet a chart 1`] = `
<div>
  <div>
    <LineChart
      EulerData={
        Array [
          Object {
            "x": 0,
            "y": 1,
          },
          Object {
            "x": 0.05,
            "y": 1.05,
          },
          Object {
            "x": 0.1,
            "y": 1.1,
          },
          Object {
            "x": 0.15,
            "y": 1.16,
          },
          Object {
            "x": 0.2,
            "y": 1.22,
          },
          Object {
            "x": 0.25,
            "y": 1.28,
          },
          Object {
            "x": 0.3,
            "y": 1.35,
          },
          Object {
            "x": 0.35,
            "y": 1.42,
          },
          Object {
            "x": 0.4,
            "y": 1.49,
          },
          Object {
            "x": 0.45,
            "y": 1.57,
          },
          Object {
            "x": 0.5,
            "y": 1.65,
          },
          Object {
            "x": 0.55,
            "y": 1.73,
          },
          Object {
            "x": 0.6,
            "y": 1.82,
          },
          Object {
            "x": 0.65,
            "y": 1.91,
          },
          Object {
            "x": 0.7,
            "y": 2.01,
          },
          Object {
            "x": 0.75,
            "y": 2.11,
          },
          Object {
            "x": 0.8,
            "y": 2.22,
          },
          Object {
            "x": 0.85,
            "y": 2.33,
          },
          Object {
            "x": 0.9,
            "y": 2.45,
          },
          Object {
            "x": 0.95,
            "y": 2.58,
          },
          Object {
            "x": 1,
            "y": 2.71,
          },
          Object {
            "x": 1.05,
            "y": 2.85,
          },
          Object {
            "x": 1.1,
            "y": 2.99,
          },
          Object {
            "x": 1.15,
            "y": 3.14,
          },
          Object {
            "x": 1.2,
            "y": 3.3,
          },
          Object {
            "x": 1.25,
            "y": 3.47,
          },
          Object {
            "x": 1.3,
            "y": 3.65,
          },
          Object {
            "x": 1.35,
            "y": 3.84,
          },
          Object {
            "x": 1.4,
            "y": 4.03,
          },
          Object {
            "x": 1.45,
            "y": 4.24,
          },
          Object {
            "x": 1.5,
            "y": 4.45,
          },
          Object {
            "x": 1.55,
            "y": 4.68,
          },
        ]
      }
      LineNames={
        Array [
          "a",
          "b",
        ]
      }
      MidpointData={
        Array [
          Object {
            "x": 0,
            "y": 2,
          },
          Object {
            "x": 0.05,
            "y": 2.1,
          },
          Object {
            "x": 0.1,
            "y": 2.21,
          },
          Object {
            "x": 0.15,
            "y": 2.32,
          },
          Object {
            "x": 0.2,
            "y": 2.44,
          },
          Object {
            "x": 0.25,
            "y": 2.57,
          },
          Object {
            "x": 0.3,
            "y": 2.7,
          },
          Object {
            "x": 0.35,
            "y": 2.83,
          },
          Object {
            "x": 0.4,
            "y": 2.98,
          },
          Object {
            "x": 0.45,
            "y": 3.13,
          },
          Object {
            "x": 0.5,
            "y": 3.29,
          },
          Object {
            "x": 0.55,
            "y": 3.46,
          },
          Object {
            "x": 0.6,
            "y": 3.64,
          },
          Object {
            "x": 0.65,
            "y": 3.82,
          },
          Object {
            "x": 0.7,
            "y": 4.02,
          },
          Object {
            "x": 0.75,
            "y": 4.22,
          },
          Object {
            "x": 0.8,
            "y": 4.44,
          },
          Object {
            "x": 0.85,
            "y": 4.66,
          },
          Object {
            "x": 0.9,
            "y": 4.9,
          },
          Object {
            "x": 0.95,
            "y": 5.15,
          },
          Object {
            "x": 1,
            "y": 5.41,
          },
          Object {
            "x": 1.05,
            "y": 5.69,
          },
          Object {
            "x": 1.1,
            "y": 5.98,
          },
          Object {
            "x": 1.15,
            "y": 6.29,
          },
          Object {
            "x": 1.2,
            "y": 6.61,
          },
          Object {
            "x": 1.25,
            "y": 6.95,
          },
          Object {
            "x": 1.3,
            "y": 7.3,
          },
          Object {
            "x": 1.35,
            "y": 7.67,
          },
          Object {
            "x": 1.4,
            "y": 8.06,
          },
          Object {
            "x": 1.45,
            "y": 8.48,
          },
          Object {
            "x": 1.5,
            "y": 8.91,
          },
          Object {
            "x": 1.55,
            "y": 9.36,
          },
        ]
      }
      axisNames={
        Array [
          "t",
          "",
        ]
      }
      horizontalAlign="top"
      verticalAlign="center"
    />
  </div>
</div>
`;
